pinot-transport/src/main/java/com/linkedin/pinot/requestHandler/BrokerRequestHandler.java

```diff
@@ -243,383 +239,396 @@ public void validateRequest(BrokerRequest brokerRequest) {
-    }
-
-    // Step 6 : Do the reduce and return
+    // Step 3: gather response from the servers.
+    Map<ServerInstance, ByteBuf> responseMap;
     try {
-      return _brokerMetrics.timeQueryPhase(request, BrokerQueryPhase.REDUCE, new Callable<BrokerResponse>() {
-        @Override
-        public BrokerResponse call() {
-          BrokerResponse returnValue = reduceService.reduceOnDataTable(request, instanceResponseMap);
-          _brokerMetrics.addMeteredQueryValue(request, BrokerMeter.DOCUMENTS_SCANNED, returnValue.getNumDocsScanned());
-          return returnValue;
-        }
-      });
+      responseMap = compositeFuture.get();
+      Map<String, Long> responseTimes = compositeFuture.getResponseTimes();
+      scatterGatherStats.setResponseTimeMillis(responseTimes);
     } catch (Exception e) {
-      // Shouldn't happen, this is only here because timeQueryPhase() can throw a checked exception, even though the nested callable can't.
-      LOGGER.error("Caught exception while processing return", e);
-      Utils.rethrowException(e);
-      throw new AssertionError("Should not reach this");
-    }
+      LOGGER.error("Caught exception while fetching responses.", e);
+      _brokerMetrics.addMeteredTableValue(tableName, BrokerMeter.RESPONSE_FETCH_EXCEPTIONS, 1);
+      return BrokerResponseFactory.getBrokerResponseWithException(responseType,
+          QueryException.getException(QueryException.BROKER_GATHER_ERROR, e));
+    }
+    _brokerMetrics.addPhaseTiming(tableName, BrokerQueryPhase.SCATTER_GATHER, System.nanoTime() - scatterStartTime);
+
+    //Step 4: deserialize the responses.
+    long deserializationStartTime = System.nanoTime();
+    Map<ServerInstance, DataTable> dataTableMap = new HashMap<>();
+    List<ProcessingException> processingExceptions = new ArrayList<>();
+    deserializeResponses(tableName, responseMap, -1, dataTableMap, processingExceptions);
+    _brokerMetrics.addPhaseTiming(tableName, BrokerQueryPhase.DESERIALIZATION,
+        System.nanoTime() - deserializationStartTime);
+
+    // Step 5: reduce (merge) the responses and create a broker response to be returned.
+    long reduceStartTime = System.nanoTime();
+    BrokerResponse brokerResponse = reduceService.reduceOnDataTable(brokerRequest, dataTableMap, _brokerMetrics);
+    _brokerMetrics.addPhaseTiming(tableName, BrokerQueryPhase.REDUCE, System.nanoTime() - reduceStartTime);
+
+    // Attach processing exceptions and return.
+    brokerResponse.setExceptions(processingExceptions);
+    return brokerResponse;
   }
 
-  private Object getDataTableFromBrokerRequestList(final BrokerRequest federatedBrokerRequest,
-      final ReduceService reduceService, final List<BrokerRequest> requests, BucketingSelection overriddenSelection,
-      final ScatterGatherStats scatterGatherStats, final long requestId)
+  /**
+   * Process multiple broker requests on multiple tables.
```

Comment:
Slightly misleading comment: This is for the same table's offline and realtime counterparts, right? Not multiple unrelated tables?